[yooplo] iP

yooplo opened the pull request:
# Regina
> If you want to make good use of your time, you've got to know what's most important and then give it all you've got - [Lee Lacocca](https://www.brainyquote.com/quotes/lee_iacocca_133508)

Meet Regina, your friendly task manager disguised as a chatbot, designed to make tracking your tasks feel like a pleasant conversation. With Regina, mundane task management transforms into a delightful experience, helping you 
- set goals
- check off completed tasks
- and organize your day in a more engaging way! 🎉 

## Installing Regina
Before you start your friendly conversation with Regina, you need to first: 
1. Download the jar file [here](https://github.com/yooplo/ip/releases/tag/A-Jar)
2. Run the jar file
3. Start adding your tasks
4. Let her take it from there! 

Not only is Regina *fool-proof* 😬 :, but she is completely **FREE**!! 🤯 

## Features
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

If you are a Java programmer too, you can use it to practice Java too. Here is the ```main``` method:
```java
    public static void main(String[] args) {
        launch(args);
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yooplo committed:
Implement greeting and exit

yooplo committed:
Implement echo functionality

yooplo committed:
Add listing functionality

yooplo committed:
Update Regina.java class

yooplo committed:
Add Task class
Add "Mark as Done" functionality

yooplo committed:
Update Task.java
Add ToDos, Events, Deadlines

yooplo committed:
Add error handling

yooplo committed:
Add delete function
Personalised the bot

yooplo committed:
Add enums usage

yooplo committed:
Add javadocs documentation

yooplo committed:
Update code style

yooplo committed:
Add file saving functionality

yooplo committed:
Add save feature

yooplo committed:
Merge branch 'branch-Level-7'

yooplo committed:
Add dates and time to DeadlinesTask

yooplo committed:
Add date and time to EventsTask

yooplo committed:
Add the command to filter tasks on a specified date

yooplo committed:
Merge branch 'branch-Level-8'

yooplo committed:
Refactor Regina class to extract required classes

yooplo committed:
Merge branch 'branch-A-MoreOOP'

yooplo committed:
Sort classes into packages

yooplo committed:
Merge branch 'branch-A-Packages'

yooplo committed:
Merge branch 'add-gradle-support' of https://github.com/yooplo/ip

yooplo committed:
Add Gradle

yooplo committed:
Add test classes for Ui and Regina

yooplo committed:
Update Test cases

yooplo committed:
Merge branch 'branch-A-JUnit'

yooplo committed:
Build jar file

yooplo committed:
Merge branch 'branch-A-Jar'

yooplo committed:
Add javadoc for recent methods and classes

yooplo committed:
Update code based of style checks

yooplo committed:
Merge branch 'branch-A-Javadoc'

yooplo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/test/java/regina/ReginaTest.java

yooplo committed:
Add the find method to TaskList and Regina class

yooplo committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/tasks/TaskList.java
#	src/test/java/regina/ReginaTest.java

yooplo committed:
Refactor code based on CheckStyle detection

yooplo committed:
Merge branch 'branch-A-CheckStyle'

yooplo committed:
Add GUI template to chatbot

yooplo committed:
Add valid responses to Regina

yooplo committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/regina/Regina.java

yooplo committed:
Add list of tasks in GUI

yooplo committed:
Include live syncing of the checkbox with the task list

yooplo committed:
Add checkbox marking of tasks functionality

yooplo committed:
Merge branch 'branch-Level-10'

yooplo committed:
Fix bug in GUI to make window resizable

estellelim commented on src/main/java/gui/MainWindow.java:
```diff
@@ -0,0 +1,77 @@
+
+    private Regina regina = new Regina();
+
+    // Imaged for the user and Regina Chatbot
```
Typo here haha

estellelim commented on src/main/java/regina/Marker.java:
```diff
@@ -0,0 +1,68 @@
+ * The regina.Marker class provides methods to mark and unmark tasks in a task list.
+ * It interacts with the tasks.TaskList to check off tasks as completed or not.
+ */
+
```
Extra blank line between the documentation block and the class.

estellelim reviewed (commented):
LGTM - Very good job on the variable and method naming, could not find any code quality issues to nitpick on!

yadobler commented on src/main/java/dateandtime/ReginaDateAndTime.java:
```diff
@@ -0,0 +1,153 @@
+     *
+     * @return A string indicating day, date and time of this instant.
+     */
+    public static String now() {
```
Would it be slightly clearer to name the method something like "getCurrentDateTime"? I can definitely foresee myself using `now` as the method name because of `LocalDate.now()`, and then a week later forget this returns a nicely formatted string instead.

yadobler commented on src/main/java/dateandtime/ReginaDateAndTime.java:
```diff
@@ -0,0 +1,153 @@
+        if (this.date.isAfter(dateAndTime.getDate())) {
+            return false;
+        }
+        if (this.date.isEqual(dateAndTime.getDate()) && this.time.isAfter(dateAndTime.getTime())) {
```
Does this return `true` if the date and time are exactly equal? (Is that intended?)

(Not a coding standard comment, but feel free to address if you want to.)

yooplo replied:
Thank you for bringing that to my attention, it should not return true haha

yadobler commented on src/main/java/dateandtime/ReginaDateAndTime.java:
```diff
@@ -0,0 +1,153 @@
+     * @param dateAndTime The ReginaDateAndTime instance to compare against.
+     * @return true if this date and time is equal to the given instance; false otherwise.
+     */
+    public boolean isEqual(ReginaDateAndTime dateAndTime) {
```
You can overload the `equals` Object method, might be useful if you intend to add hashmaps next time! (You can also go a step further and implement a `Comparable<T>` so that you can sort by `ReginaDateAndTime`). 

(Not a code style comment, just a general coding comment)

yadobler commented on src/main/java/dateandtime/ReginaDateAndTime.java:
```diff
@@ -0,0 +1,153 @@
+        return this.time.format(DateTimeFormatter.ofPattern(OUTPUT_TIME_PATTERN));
+    }
+
+    public String toSavedFormatting() {
```
Is this a getter method or a normal method missing a Javadoc? (sorry, the method name makes me thing there's something happening, but there's no Javadoc in that sense)

yadobler commented on src/main/java/file/Storage.java:
```diff
@@ -0,0 +1,101 @@
+        } else if (taskType.equals("D")) {
+            String taskDeadline = taskParts[3];
+            task = new DeadlinesTask(taskDescription, taskDeadline);
+        } else { // taskType is "E"
```
Thank you for adding comments throughout to explain the thought process, definitely something many won't bother. Helps with following the logic!

yadobler commented on src/main/java/regina/Parser.java:
```diff
@@ -0,0 +1,47 @@
+     * @return True if valid, false otherwise.
+     */
+    private boolean isValidCommand(String command) {
+        return command.equals("help") || command.equals("now")
```
I'm not sure if coding standard would prefer all the `||`s to be newline. I know the _block-like construct_ for array initialisation prefers these:

```java
new int[] {           new int[] {
  0, 1, 2, 3            0,
}                       1,
                        2,
new int[] {             3,
  0, 1,               }
  2, 3
}   
```

Personally I prefer all new-line, or the second `||` to be all aligned. What do you feel?

yadobler commented on src/main/java/regina/Ui.java:
```diff
@@ -0,0 +1,148 @@
+ * It provides methods to display messages, prompts, and help information to the user.
+ * The class manages the input and output operations, ensuring a user-friendly command-line interface.
+ *
+ * <p>This class includes the following functionalities:
```
Great use of html formatting in JavaDoc!

yadobler reviewed (commented):
LGTM! Just a few nits on how I interpret the coding style should be, but nothing glaringly wrong. 😄

yooplo committed:
Allow chat segment to follow true to window size

yooplo committed:
Merge branch 'branch-Level-10'

yooplo committed:
Refactor code
Some methods are rather long.

Shorter methods make it easier to follow the logic of the code.

Let's,
* extract some inline methods
* extract variables for better readability

yooplo committed:
Add assertions
Some methods allow for null values to be executed in the code.

Null values are not expected as the chatbot is supposed to take non-empty inputs from the user. Poor usage of the methods may give unwanted results.

Let's
* assert that arguments are not null in the relevant methods

Done this way to ensure the intended behaviour by the chatbot regardless of the user's inputs.

yooplo committed:
Revert "Add assertions"

This reverts commit acaa1318c10980113b7b3939760b6b11e38e6187.

yooplo committed:
Reapply "Add assertions"

This reverts commit 2ad48c7e90b1f24cca82baf0cb781bb3cb878343.

yooplo committed:
Revert "Reapply "Add assertions""

This reverts commit f2ca7df2047c489f43662b96dad179a5a38ace3c.

yooplo committed:
Merge pull request #2 from yooplo/master

Add assertions
Some methods allow for null values to be executed in the code.

Null values are not expected as the chatbot is supposed to take non-empty inputs from the user. Poor usage of the methods may give unwanted results.

Let's
* assert that arguments are not null in the relevant methods

Done this way to ensure the intended behaviour by the chatbot regardless of the user's inputs.

yooplo committed:
Merge pull request #1 from yooplo/branch-A-CodeQuality

Refactor code

yooplo committed:
Merge branch 'master' of https://github.com/yooplo/ip

yooplo committed:
Merge branch 'branch-A-Assertions' of https://github.com/yooplo/ip

yooplo committed:
Add assertions
Some methods allow for null values to be executed in the code.

Null values are not expected as the chatbot is supposed to take non-empty inputs from the user. Poor usage of the methods may give unwanted results.

Let's
* assert that arguments are not null in the relevant methods

Done this way to ensure the intended behaviour by the chatbot regardless of the user's inputs.

yooplo committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/regina/Parser.java

yooplo committed:
Refactor some methods to include streams

yooplo committed:
Add snooze for tasks with deadlines

Snoozing without specifying values will snooze the task by 30 minutes on default.
Updated snoozed deadline is resaved into the saved data file.

yooplo committed:
Merge pull request #3 from yooplo/branch-A-Streams

Refactor some methods to include streams

yooplo committed:
Merge branch 'branch-A-Streams' of https://github.com/yooplo/ip into branch-B-Snooze

# Conflicts:
#	data/savedData.txt
#	src/main/java/regina/Regina.java

yooplo committed:
Merge pull request #4 from yooplo/branch-B-Snooze

Add snooze for tasks with deadlines

yooplo committed:
Add numbering to list of task in the checkbox column of GUI

yooplo committed:
Bug Fixes

Fix big where if in minutes, the task is always snoozed by the default value of 30 minutes.

yooplo committed:
Merge branch 'branch-B-Snooze'

yooplo committed:
Refactor code by extracting methods out

yooplo committed:
Add test case for snooze function

yooplo committed:
Add test cases for ReginaDateAndTime

yooplo committed:
Correct the expected output of testGreet() in the UITest class

yooplo committed:
Merge pull request #6 from yooplo/A-MoreTesting

Add test cases for ReginaDateAndTime

yooplo committed:
Set up product website

yooplo committed:
Update README.md

yooplo committed:
Update README.md

yooplo committed:
Update README.md

yooplo committed:
Update build.gradle

yooplo committed:
Rename UI.png.png to UI.png

yooplo committed:
Rename UI.png to Ui.png

yooplo committed:
Update README.md

yooplo committed:
Add 'bye' command

yooplo committed:
Merge branch 'branch-A-Bye'

# Conflicts:
#	docs/README.md

yooplo committed:
Update README.md
